core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferServiceSuite.scala

```diff
@@ -77,6 +87,68 @@ class NettyBlockTransferServiceSuite
     verifyServicePort(expectedPort = service0.port + 1, actualPort = service1.port)
   }
 
+  test("SPARK-27637: test fetch block with executor dead") {
+    implicit val exectionContext = ExecutionContext.global
+    val port = 17634 + Random.nextInt(10000)
+    logInfo("random port for test: " + port)
+
+    val driverEndpointRef = new RpcEndpointRef(new SparkConf()) {
+      override def address: RpcAddress = null
+      override def name: String = "test"
+      override def send(message: Any): Unit = {}
+      // This rpcEndPointRef always return false for unit test to touch ExecutorDeadException.
+      override def ask[T: ClassTag](message: Any, timeout: RpcTimeout): Future[T] = {
+        Future{false.asInstanceOf[T]}
+      }
+    }
+
+    val clientFactory = mock(classOf[TransportClientFactory])
+    val client = mock(classOf[TransportClient])
+    // This is used to touch an IOException during fetching block.
+    when(client.sendRpc(any(), any())).thenAnswer(_ => {throw new IOException()})
+    when(clientFactory.createClient(any(), any())).thenReturn(client)
+
+    val listener = mock(classOf[BlockFetchingListener])
+    // This is used to throw a ExecutorDeadException for unit test, when the fetch is failed
+    // because of ExecutorDeadException.
+    when(listener.onBlockFetchFailure(any(), any(classOf[ExecutorDeadException])))
+      .thenAnswer(_ => {throw new ExecutorDeadException("Executor is dead.")})
+
+    service0 = createService(port, driverEndpointRef)
+    val clientFactoryField = service0.getClass.getField(
+      "org$apache$spark$network$netty$NettyBlockTransferService$$clientFactory")
+    clientFactoryField.setAccessible(true)
+    clientFactoryField.set(service0, clientFactory)
+
+    // The ExecutorDeadException is thrown by listener when fetch failed for ExecutorDeadException.
+    intercept[ExecutorDeadException](service0.fetchBlocks("localhost", port, "exec1",
+      Array("block1"), listener, mock(classOf[DownloadFileManager])))
+
+    val blockFetchStarter = new RetryingBlockFetcher.BlockFetchStarter {
+      override def createAndStart(blockIds: Array[String], listener: BlockFetchingListener) {
+        try {
+          new OneForOneBlockFetcher(client, "appId", "execId", blockIds, listener,
+            mock(classOf[TransportConf]), mock(classOf[DownloadFileManager])).start()
+        } catch {
+          case e: IOException =>
+            if (driverEndpointRef.askSync[Boolean](IsExecutorAlive("execId"))) {
```

Comment:
This is the code we want to test, we should not put it in the test...

Reply:
one principal to write a regression test: it should fail without the patch

Reply:
Get it, I wanted to prove that the retryingBlockFetch would not retry when the relative executor is dead. I have remove this part.